src/git.rs

```diff
@@ -444,3 +494,171 @@ impl Drop for TempRepo {
+        assert_eq!(git.is_mainline("refs/heads/tagged-branch").unwrap(), true);
+    }
+
+    #[test]
+    fn is_mainline_with_literal_config_does_not_identify_head_branch() {
+        let project_root = env!("CARGO_MANIFEST_DIR");
+        let git = LiveGit::new(
+            format!("{}/tests/test-repo.git", project_root),
+            GitConfig {
+                mainlines: vec!["tagged-branch".into()],
+            },
+        )
+        .unwrap();
+        assert_eq!(git.is_mainline("refs/heads/master").unwrap(), false);
+        assert_eq!(git.is_mainline("refs/heads/tagged-branch").unwrap(), true);
+    }
+
+    #[quickcheck]
+    fn is_mainline_fuzz(branch: String, mainlines: Vec<String>) {
+        if valid_mainlines(&mainlines) {
+            let project_root = env!("CARGO_MANIFEST_DIR");
+            let git = LiveGit::new(
+                format!("{}/tests/test-repo.git", project_root),
+                GitConfig { mainlines },
+            )
+            .unwrap();
+            git.is_mainline(&branch).unwrap();
+        }
+    }
+
+    #[test]
+    fn is_mainline_with_multiple_glob_config_identifies_all_matches() {
+        let project_root = env!("CARGO_MANIFEST_DIR");
+        let git = LiveGit::new(
+            format!("{}/tests/test-repo.git", project_root),
+            GitConfig {
+                mainlines: vec!["HEAD".into(), "tagged-*".into()],
+            },
+        )
+        .unwrap();
+        assert_eq!(git.is_mainline("refs/heads/master").unwrap(), true);
+        assert_eq!(git.is_mainline("refs/heads/tagged-branch").unwrap(), true);
+    }
+
+    #[test]
+    fn new_commits_off_master_with_default_config() {
+        let project_root = env!("CARGO_MANIFEST_DIR");
+        let git = LiveGit::default(format!("{}/tests/test-repo.git", project_root)).unwrap();
+        let commits = git
+            .find_new_commits(
+                &[Oid::from_str("eb5e0185546b0bb1a13feec6b9ee8b39985fea42").unwrap()],
+                &[Oid::from_str("6004dfdb071c71e5e76ad55b924b576487e1c485").unwrap()],
+                &None,
+            )
+            .unwrap();
+        assert_eq!(commits.len(), 2)
+    }
+
+    #[test]
+    fn new_commits_off_master_with_configured_mainline_glob() {
+        let project_root = env!("CARGO_MANIFEST_DIR");
+        let git = LiveGit::new(
+            format!("{}/tests/test-repo.git", project_root),
+            GitConfig {
+                mainlines: vec!["HEAD".into(), "valid-*".into()],
+            },
+        )
+        .unwrap();
+        let commits = git
+            .find_new_commits(
+                &[Oid::from_str("eb5e0185546b0bb1a13feec6b9ee8b39985fea42").unwrap()],
+                &[Oid::from_str("6004dfdb071c71e5e76ad55b924b576487e1c485").unwrap()],
+                &None,
+            )
+            .unwrap();
+        assert_eq!(commits.len(), 1)
+    }
+
+    #[test]
+    fn new_commits_off_master_with_configured_mainline_literal_branch_name() {
```

Comment:
I feel like this test name isn't telling me enough about the test.  For instance I don't know why we only end up with one commit in the end.  Was it because the second mainline had one of the commits, or the first?
Perhaps something like, finds_the_least_commits_ahead_of_any_mainline_branch?
Also; what happens when you have multiple mainlines with divergent histories?

Reply:
Agree about of the name of the test.
I've also been thinking that the test body relying on the state of a checked in repo is also hard, because it's hidden state. One option might be to add a git-log of the relevant commits as a comment in the test.